ephios/plugins/federation/views/frontend.py

```diff
@@ -0,0 +1,169 @@
+from django.urls import reverse, reverse_lazy
+from django.utils.translation import gettext_lazy as _
+from django.views.generic import CreateView, DeleteView, DetailView, FormView, TemplateView
+from guardian.mixins import LoginRequiredMixin
+from requests import HTTPError, JSONDecodeError, ReadTimeout
+from rest_framework.exceptions import PermissionDenied
+
+from ephios.core.models import Event
+from ephios.core.views.signup import BaseShiftActionView
+from ephios.extra.mixins import StaffRequiredMixin
+from ephios.plugins.federation.forms import InviteCodeForm, RedeemInviteCodeForm
+from ephios.plugins.federation.models import (
+    FederatedEventShare,
+    FederatedGuest,
+    FederatedHost,
+    FederatedUser,
+    InviteCode,
+)
+from ephios.plugins.federation.views.api import FederationOAuthView
+
+
+class ExternalEventListView(LoginRequiredMixin, TemplateView):
+    template_name = "federation/external_event_list.html"
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        events = []
+        for host in FederatedHost.objects.all():
+            try:
+                r = requests.get(
+                    urljoin(host.url, reverse("federation:shared_event_list_view")),
+                    headers={"Authorization": f"Bearer {host.access_token}"},
+                    timeout=5,
+                )
+                r.raise_for_status()
+                results = r.json()["results"]
+                for event in results:
+                    event["type"] = event["type"]["title"]
+                    event["start_time"] = datetime.fromisoformat(event["start_time"])
+                    event["end_time"] = datetime.fromisoformat(event["end_time"])
+                    event["host"] = host.name
+                events += results
+            except (HTTPError, JSONDecodeError, ReadTimeout):
+                continue
+        events.sort(key=lambda e: e["start_time"])
+        context["events"] = events
+        return context
+
+
+class CheckFederatedAccessTokenMixin:
+    def dispatch(self, request, *args, **kwargs):
+        if "access_token" not in request.session.keys():
+            return FederationOAuthView.as_view()(request, *args, **kwargs)
+        return super().dispatch(request, *args, **kwargs)
+
+    def get_context_data(self, object):
+        context = super().get_context_data()
+        context["guest_url"] = FederatedGuest.objects.get(pk=self.request.session["guest"]).url
+        return context
+
+
+class FederatedEventDetailView(CheckFederatedAccessTokenMixin, DetailView):
+    model = Event
+    template_name = "federation/event_detail.html"
+
+    def get_object(self, queryset=None):
+        obj = super().get_object(queryset)
+        try:
+            guest = self.request.session["guest"]
+            FederatedEventShare.objects.get(
+                event=obj,
+                shared_with__in=[FederatedGuest.objects.get(pk=guest)],
```

Comment:
I get a `FederatedGuest.DoesNotExist` server error after recreating the connection and accessing the share again. Possibly because `guest` contains the old `FederatedGuest` primary key? Let me know if you need more help recreating this issue.

Reply:
what do you mean by "recreating the connection"?

Reply:
Removing the federation and then recreating the federation

Reply:
I think that's impossible to prevent as we have no way of invalidating this session when the connection is removed. I added a `session.flush()` here, so the user would be presented with a permission denied, but on the next try everything would be set up correctly again

Reply:
Of course the user has to give permission again - fine, but the `DoesNotExist` is preventable...